File: tools/metilene_input_plus.py
#!/usr/bin/env python3
"""
Prepare metilene input matrix from two groups of sorted bedGraph files.
Supports parallel processing by chromosome.
Automatically skips chromosomes: chrUn_*, *_random, *_alt, *_fix.
Filtering rule: keep a site only if BOTH groups have non-missing ratio >= --ratio.
"""

import argparse
import os
import sys
import subprocess
import tempfile
import shutil
import re
from pathlib import Path
from concurrent.futures import ProcessPoolExecutor, as_completed
import gzip

def pretty_time():
    from datetime import datetime
    return datetime.now().strftime("%a %b %d, %H:%M:%S, %Y")


def log(msg, level="INFO"):
    print(f"[{level}]\t{pretty_time()}\t{msg}", file=sys.stderr)


def parse_args():
    parser = argparse.ArgumentParser(
        description="Generate metilene input from two groups of sorted bedGraph files.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Notes:
  - Input bedGraph files MUST be sorted (bedtools sort -i file.bg > file.sorted.bg)
  - Prefer --in1_list / --in2_list when you have many samples
  - Chromosomes matching chrUn_*, *_random, *_alt, *_fix are automatically skipped
  - A site is kept only if BOTH group1 and group2 have non-missing ratio >= --ratio
        """
    )

    # Input
    parser.add_argument("--in1", help="Comma-separated group1 bedGraph files")
    parser.add_argument("--in2", help="Comma-separated group2 bedGraph files")
    parser.add_argument("--in1_list", help="File with group1 paths (one per line)")
    parser.add_argument("--in2_list", help="File with group2 paths (one per line)")

    # Output & groups
    parser.add_argument("--out", help="Output file (default: metilene_<h1>_<h2>.input)")
    parser.add_argument("--h1", default="g1", help="Group1 name (default: g1)")
    parser.add_argument("--h2", default="g2", help="Group2 name (default: g2)")

    # Filtering & performance
    parser.add_argument("--ratio", type=float, default=0.0,
                        help="Min non-missing ratio required in EACH group (0-1, default: 0)")
    parser.add_argument("--NA", default=".", help="Missing value symbol (default: .)")
    parser.add_argument("-t", "--threads", type=int, default=4,
                        help="Parallel jobs (default: 4; set 1 for single-pass)")
    parser.add_argument("-b", "--bedtools", default="bedtools",
                        help="Path to bedtools")
    parser.add_argument("--tmpdir", default="./tmp/",
                        help="Directory for temporary files (default: system temp)")

    args = parser.parse_args()

    if not ((args.in1 or args.in1_list) and (args.in2 or args.in2_list)):
        parser.error("Must provide --in1/--in2 or --in1_list/--in2_list")
    if not (0 <= args.ratio <= 1):
        parser.error("--ratio must be between 0 and 1")
    if args.threads < 1:
        parser.error("--threads must be >= 1")

    return args


def read_file_list(path):
    files = []
    with open(path) as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith("#"):
                files.append(line)
    if not files:
        raise ValueError(f"Empty file list: {path}")
    return files


def check_files(files):
    resolved = []
    for f in files:
        p = Path(f).resolve()
        if not p.exists():
            raise FileNotFoundError(f"File not found: {p}")
        if not os.access(p, os.R_OK):
            raise PermissionError(f"File not readable: {p}")
        resolved.append(str(p))
    return resolved


def is_valid_chrom(chrom):
    """Skip chrUn_*, *_random, *_alt, *_fix."""
    if chrom.startswith("chrUn_"):
        return False
    if chrom.endswith(("_random", "_alt", "_fix")):
        return False
    return True


def get_chromosomes(bedgraph):
    chroms = set()
    with gzip.open(bedgraph, "rt") as f:
        for line in f:
            if line.startswith(("#", "track", "browser")):
                continue
            chrom = line.split("\t", 1)[0]
            if chrom and is_valid_chrom(chrom):
                chroms.add(chrom)

    def key(c):
        m = re.search(r"(\d+)", c)
        return (0, int(m.group(1)), c) if m else (1, 0, c)

    return sorted(chroms, key=key)


def run_unionbedg(bedtools, names, na, file_list, awk_filter, outfile):
    """Run bedtools unionbedg + cut/sed/awk pipeline."""
    with tempfile.NamedTemporaryFile("w", delete=False, suffix=".txt") as lf:
        lf.write("\n".join(file_list) + "\n")
        listfile = lf.name

    cmd = (
        f"{bedtools} unionbedg -header -names {names} -filler '{na}' "
        f"-i $(cat {listfile}) "
        f"| cut -f1,3- "
        f"| sed 's/end/pos/' "
        f"| awk '{awk_filter}' "
        f"> {outfile}"
    )
    try:
        subprocess.run(cmd, shell=True, check=True, executable="/bin/bash")
    finally:
        os.unlink(listfile)


def filter_to_chrom(src, dest, chrom):
    """Extract lines of one chromosome."""
    with gzip.open(src, "rt") as fin, gzip.open(dest, "wt") as fout:
        for line in fin:
            if line.startswith(("#", "track", "browser")):
                continue
            if line.startswith(chrom + "\t"):
                fout.write(line)


def filter_valid_chroms(src, dest, valid_chroms):
    """Keep only lines from valid chromosomes."""
    with gzip.open(src, "rt") as fin, gzip.open(dest, "wt") as fout:
        for line in fin:
            if line.startswith(("#", "track", "browser")):
                continue
            chrom = line.split("\t", 1)[0]
            if chrom in valid_chroms:
                fout.write(line)


def process_one_chrom(chrom, short_paths, bedtools, names, na, awk_filter, out_dir):
    """Worker: process a single chromosome."""
    work_dir = Path(tempfile.mkdtemp(prefix=f"chr_{chrom}_", dir=out_dir))
    try:
        chr_files = []
        for i, src in enumerate(short_paths):
            dest = work_dir / f"{i}.bg"
            filter_to_chrom(src, dest, chrom)
            chr_files.append(str(dest))

        chr_out = Path(out_dir) / f"{chrom}.out"
        run_unionbedg(bedtools, names, na, chr_files, awk_filter, str(chr_out))
        return chrom, str(chr_out)
    finally:
        shutil.rmtree(work_dir, ignore_errors=True)


def main():
    args = parse_args()

    log("Input files must be sorted (bedtools sort)", "WARNING")
    log("Checking input...")

    # Load groups
    g1 = read_file_list(args.in1_list) if args.in1_list else \
         [x.strip() for x in args.in1.split(",") if x.strip()]
    g2 = read_file_list(args.in2_list) if args.in2_list else \
         [x.strip() for x in args.in2.split(",") if x.strip()]

    g1 = check_files(g1)
    g2 = check_files(g2)

    n1 = len(g1)
    n2 = len(g2)

    names = " ".join([args.h1] * n1 + [args.h2] * n2)
    out_file = str(Path(args.out or f"metilene_{args.h1}_{args.h2}.input").resolve())

    if shutil.which(args.bedtools) is None and not Path(args.bedtools).exists():
        sys.exit(f"ERROR: bedtools not found: {args.bedtools}")

    log(f"group1 = {n1}, group2 = {n2}")
    log(f"ratio (per group) = {args.ratio}, threads = {args.threads}")
    log(f"Output: {out_file}")

    # Awk filter: keep site only if BOTH groups meet the non-missing ratio
    # After cut -f1,3- : $1=chrom, $2=pos, $3..(2+n1)=group1, then group2
    awk_filter = (
        f'NR==1{{print; next}} '
        f'{{ '
        f'  n1={n1}; n2={n2}; '
        f'  miss1=0; miss2=0; '
        f'  for(i=3; i<=2+n1; i++) if($i=="{args.NA}") miss1++; '
        f'  for(i=3+n1; i<=2+n1+n2; i++) if($i=="{args.NA}") miss2++; '
        f'  if( (n1-miss1)/n1 >= {args.ratio} && (n2-miss2)/n2 >= {args.ratio} ) print; '
        f'}}'
    )

    # Temporary directory
    if args.tmpdir:
        root = Path(args.tmpdir).resolve()
        root.mkdir(parents=True, exist_ok=True)
        tmpdir = Path(tempfile.mkdtemp(prefix="metilene_", dir=root))
        need_cleanup = True
    else:
        tmp_ctx = tempfile.TemporaryDirectory(prefix="metilene_")
        tmpdir = Path(tmp_ctx.name)
        need_cleanup = False

    try:
        log(f"Temp dir: {tmpdir}")

        # Short symlinks
        short_paths = []
        for i, f in enumerate(g1 + g2):
            link = tmpdir / f"{i}.bg"
            link.symlink_to(f)
            short_paths.append(str(link))

        # Always get valid chromosomes
        log("Collecting valid chromosomes (skip chrUn_*/_random/_alt/_fix)...")
        chroms = get_chromosomes(short_paths[0])
        valid_set = set(chroms)
        log(f"Valid chromosomes: {len(chroms)}")

        if args.threads == 1:
            # Single-pass with chromosome filter
            log("Single-pass mode (with chromosome filter)...")
            filtered = []
            for i, src in enumerate(short_paths):
                dest = tmpdir / f"filt_{i}.bg"
                filter_valid_chroms(src, dest, valid_set)
                filtered.append(str(dest))
            run_unionbedg(args.bedtools, names, args.NA, filtered, awk_filter, out_file)

        else:
            # Parallel by chromosome
            log(f"Parallel mode ({args.threads} threads)...")
            chr_dir = tmpdir / "chr_out"
            chr_dir.mkdir()

            results = {}
            with ProcessPoolExecutor(max_workers=args.threads) as pool:
                futures = {
                    pool.submit(process_one_chrom, chrom, short_paths,
                                args.bedtools, names, args.NA, awk_filter, str(chr_dir)): chrom
                    for chrom in chroms
                }
                for fut in as_completed(futures):
                    chrom, path = fut.result()
                    results[chrom] = path
                    log(f"Done: {chrom}")

            # Merge in order, keep one header
            log("Merging results...")
            header_done = False
            with open(out_file, "w") as out:
                for chrom in chroms:
                    path = results.get(chrom)
                    if not path or not Path(path).exists():
                        continue
                    with open(path) as fh:
                        for line in fh:
                            if line.startswith(("chrom\t", "chr\t")):
                                if not header_done:
                                    out.write(line)
                                    header_done = True
                                continue
                            out.write(line)

    finally:
        if need_cleanup and tmpdir.exists():
            shutil.rmtree(tmpdir, ignore_errors=True)

    print("\n*****\n", file=sys.stderr)
    print(f"[BASIC CALL] metilene -t 4 -a {args.h1} -b {args.h2} {out_file} > out.file",
          file=sys.stderr)
    print("Adjust -t / -a / -b as needed.\n", file=sys.stderr)


if __name__ == "__main__":
    main()

File: tools/metilene_input_plus.pl
#!/usr/bin/perl

use strict;
use warnings;

use Getopt::Long;
use File::Spec;
use File::Basename;
use File::Temp qw/tempdir tempfile/;

# -----------------------------------------------------------------------------
# Variables

my ($USAGE, $call, $ret, $help);
my ($in1, $in2, $in1_list, $in2_list, $out, $h1, $h2, $bedtools);
my ($NA, $out_metilene, $g1_h, $g2_h);
my ($ratio);
my (@g1, @g2);
my $SCRIPTNAME = basename($0);

# -----------------------------------------------------------------------------
# OPTIONS

$USAGE = << "USE";

    usage:  perl $SCRIPTNAME [options]

    === 输入文件（二选一） ===
        --in1 <list>        逗号分隔的 group1 bedGraph 文件（样本少时用）
        --in2 <list>        逗号分隔的 group2 bedGraph 文件（样本少时用）

        --in1_list <file>   group1 文件列表（每行一个路径，推荐！样本多时用这个）
        --in2_list <file>   group2 文件列表（每行一个路径，推荐！样本多时用这个）

    === 其他参数 ===
        --out <file>        输出文件（默认: metilene_<h1>_<h2>.input）
        --h1 <string>       group1 名称（默认: g1）
        --h2 <string>       group2 名称（默认: g2）
        --ratio <float>     保留位点的最小非缺失比例（0~1，默认: 0，即不过滤）
                            例: 0.75 表示至少 75% 样本有值才保留该位点
        --NA <string>       缺失值符号（默认: .）
        -b <path>           bedtools 路径（默认从 PATH 查找）
        -h|--help           显示帮助

    注意: 输入 bedGraph 必须已经排序！
          bedtools sort -i file.bg > file.sorted.bg

USE

if (!@ARGV) {
    printf STDERR $USAGE;
    exit -1;
}

unless (GetOptions(
    "in1=s"       => \$in1,
    "in2=s"       => \$in2,
    "in1_list=s"  => \$in1_list,
    "in2_list=s"  => \$in2_list,
    "out=s"       => \$out,
    "h1=s"        => \$h1,
    "h2=s"        => \$h2,
    "ratio=f"     => \$ratio,
    "NA=s"        => \$NA,
    "b=s"         => \$bedtools,
    "h|help"      => \$help
)){
    printf STDERR $USAGE;
    exit -1;
}
if (defined $help){
    printf STDERR $USAGE;
    exit -1;
}

# -----------------------------------------------------------------------------
# MAIN

############
## checks ##
############
print STDERR ("[WARNING]" . prettyTime() . "Input files need to be SORTED, i.e., use \"bedtools sort -i file >file.sorted\"\n");
print STDERR ("[INFO]" . prettyTime() . "Checking flags\n");

# ----- group 1 -----
if (defined $in1_list) {
    # 从文件列表读取（推荐，避免 ARG_MAX）
    $in1_list = File::Spec->rel2abs($in1_list);
    die "##### AN ERROR: --in1_list file not found: $in1_list\n" unless -e $in1_list;
    die "##### AN ERROR: --in1_list file not readable: $in1_list\n" unless -r $in1_list;

    open(my $fh, "<", $in1_list) or die "##### Cannot open $in1_list: $!\n";
    while (my $line = <$fh>) {
        chomp $line;
        $line =~ s/^\s+|\s+$//g;          # trim
        next if $line eq "" || $line =~ /^#/;  # skip empty / comment
        push @g1, $line;
    }
    close $fh;
    die "##### AN ERROR: --in1_list is empty\n" if scalar(@g1) == 0;
}
elsif (defined $in1) {
    @g1 = split(/,/, $in1);
}
else {
    die "##### AN ERROR: must provide either --in1 or --in1_list\n";
}

for (my $i = 0; $i < scalar(@g1); $i++){
    $g1[$i] = File::Spec->rel2abs($g1[$i]);
    if (-e $g1[$i]) {
        die "##### AN ERROR: $g1[$i] not readable\n" unless -r $g1[$i];
    } else {
        die "##### AN ERROR: file not found: $g1[$i]\n";
    }
}

# ----- group 2 -----
if (defined $in2_list) {
    $in2_list = File::Spec->rel2abs($in2_list);
    die "##### AN ERROR: --in2_list file not found: $in2_list\n" unless -e $in2_list;
    die "##### AN ERROR: --in2_list file not readable: $in2_list\n" unless -r $in2_list;

    open(my $fh, "<", $in2_list) or die "##### Cannot open $in2_list: $!\n";
    while (my $line = <$fh>) {
        chomp $line;
        $line =~ s/^\s+|\s+$//g;
        next if $line eq "" || $line =~ /^#/;
        push @g2, $line;
    }
    close $fh;
    die "##### AN ERROR: --in2_list is empty\n" if scalar(@g2) == 0;
}
elsif (defined $in2) {
    @g2 = split(/,/, $in2);
}
else {
    die "##### AN ERROR: must provide either --in2 or --in2_list\n";
}

for (my $i = 0; $i < scalar(@g2); $i++){
    $g2[$i] = File::Spec->rel2abs($g2[$i]);
    if (-e $g2[$i]) {
        die "##### AN ERROR: $g2[$i] not readable\n" unless -r $g2[$i];
    } else {
        die "##### AN ERROR: file not found: $g2[$i]\n";
    }
}

# ----- defaults -----
if (!defined $h1) { $h1 = "g1"; }
if (!defined $h2) { $h2 = "g2"; }
if (!defined $NA) { $NA = "."; }
if (!defined $ratio) { $ratio = 0; }

if ($ratio < 0 || $ratio > 1) {
    die "##### AN ERROR: --ratio must be between 0 and 1 (got $ratio)\n";
}

# build header name strings
$g1_h = $h1;
$g2_h = $h2;
for (my $i = 1; $i < scalar(@g1); $i++){ $g1_h .= " $h1"; }
for (my $i = 1; $i < scalar(@g2); $i++){ $g2_h .= " $h2"; }

# output file
if (defined $out){
    $out_metilene = File::Spec->rel2abs($out);
} else {
    $out_metilene = "metilene_${h1}_${h2}.input";
}

## bedtools ##
if (defined $bedtools){
    $bedtools = File::Spec->rel2abs($bedtools);
    die "##### AN ERROR: bedtools not found: $bedtools\n" unless -e $bedtools;
    die "##### AN ERROR: bedtools is a directory\n" if -d $bedtools;
    die "##### AN ERROR: bedtools not executable\n" unless -x $bedtools;
} else {
    $bedtools = "bedtools";
}
$ret = system("command -v $bedtools &> /dev/null");
if ($ret != 0){
    die "##### AN ERROR: No bedtools found. Use -b to specify path.\n";
}

#####################
## union + filter  ##
#####################
my $total_samples = scalar(@g1) + scalar(@g2);
print STDERR ("[INFO]" . prettyTime() . "group1 = " . scalar(@g1) . " samples, group2 = " . scalar(@g2) . " samples\n");
print STDERR ("[INFO]" . prettyTime() . "Total samples = $total_samples  |  ratio threshold = $ratio\n");
print STDERR ("[INFO]" . prettyTime() . "Write metilene input to $out_metilene\n");

# Create short symlinks + file list to avoid ARG_MAX inside bedtools
my $tmpdir = tempdir(CLEANUP => 1);
print STDERR ("[INFO]" . prettyTime() . "Temporary directory: $tmpdir\n");

my @all_files = (@g1, @g2);
my @short_paths;
for (my $i = 0; $i < scalar(@all_files); $i++) {
    my $link = File::Spec->catfile($tmpdir, "$i.bg");
    symlink($all_files[$i], $link)
        or die "##### Cannot create symlink $link -> $all_files[$i]\n";
    push @short_paths, $link;
}

my $listfile = File::Spec->catfile($tmpdir, "filelist.txt");
open(my $lfh, ">", $listfile) or die "##### Cannot write $listfile: $!\n";
print $lfh join("\n", @short_paths), "\n";
close $lfh;

# awk filter
my $awk_code = 'NR==1{print;next}{non_miss=0;for(i=3;i<=NF;i++)if($i!="'.$NA.'")non_miss++;if(non_miss/(NF-2)>='.$ratio.')print}';

# Write pipeline to a temporary shell script (command line stays short)
my ($script_fh, $script_path) = tempfile(
    "metilene_unionXXXXXX",
    DIR    => $tmpdir,
    SUFFIX => ".sh",
    UNLINK => 1
);

print $script_fh "#!/bin/bash\n";
print $script_fh "set -euo pipefail\n\n";
print $script_fh "mapfile -t FILES < '$listfile'\n";
print $script_fh "$bedtools unionbedg -header -names $g1_h $g2_h -filler '$NA' -i \"\${FILES[@]}\" \\\n";
print $script_fh "  | cut -f1,3- \\\n";
print $script_fh "  | sed 's/end/pos/' \\\n";
print $script_fh "  | awk '$awk_code' \\\n";
print $script_fh "  > '$out_metilene'\n";

close $script_fh;
chmod 0755, $script_path;

print STDERR ("[INFO]" . prettyTime() . "Running unionbedg pipeline...\n");
$ret = system($script_path);
if ($ret != 0) {
    die "##### AN ERROR while running unionbedg pipeline (exit $ret)\n";
}

print STDERR ("\n*****\n\n"
    . "[BASIC CALL:] metilene -t 4 -a $h1 -b $h2 $out_metilene > out.file\n"
    . "Please adjust threads (-t) and group names (-a/-b) as needed.\n");


# -----------------------------------------------------------------------------
# FUNCTIONS

sub prettyTime {
    my @months   = qw(Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec);
    my @weekDays = qw(Sun Mon Tue Wed Thu Fri Sat Sun);
    my ($second, $minute, $hour, $dayOfMonth, $month, $yearOffset, $dayOfWeek, $dayOfYear, $daylightSavings) = localtime();
    my $year = 1900 + $yearOffset;
    return "\t$weekDays[$dayOfWeek] $months[$month] $dayOfMonth, $hour:$minute:$second, $year\t";
}